File: frontend/app/services/generativeUI.ts
export interface UIComponent {
  type: string;
  props: Record<string, any>;
  children?: string | UIComponent[];
  styles?: Record<string, string>;
}

export class GenerativeUIService {
  private baseUrl: string;

  constructor(baseUrl: string = '/api/ai') {
    this.baseUrl = baseUrl;
  }

  async generateComponent(description: string, componentType: string): Promise<UIComponent> {
    try {
      const response = await fetch(`${this.baseUrl}/generate-ui`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          description,
          componentType,
          context: 'ekioba-marketplace'
        }),
      });

      if (!response.ok) {
        throw new Error(`Failed to generate UI: ${response.statusText}`);
      }

      const data = await response.json();
      return data.component;
    } catch (error) {
      console.error('Error generating UI component:', error);
      // Fallback to basic component
      return {
        type: componentType,
        props: {
          className: 'generated-component bg-blue-500 text-white px-4 py-2 rounded',
          children: description,
        },
      };
    }
  }

  async generateLayout(description: string): Promise<UIComponent[]> {
    try {
      const response = await fetch(`${this.baseUrl}/generate-layout`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          description,
          context: 'ekioba-marketplace'
        }),
      });

      if (!response.ok) {
        throw new Error(`Failed to generate layout: ${response.statusText}`);
      }

      const data = await response.json();
      return data.layout;
    } catch (error) {
      console.error('Error generating layout:', error);
      return [];
    }
  }
}

export const generativeUIService = new GenerativeUIService();

File: my-project/backend/app/api/generative_ui.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from fastapi import APIRouter
from pydantic import BaseModel

router = APIRouter()


class UIComponent(BaseModel):
    type: str
    props: Dict[str, Any]
    children: Optional[List["UIComponent"]] = None
    styles: Optional[Dict[str, str]] = None


class GenerateUIRequest(BaseModel):
    description: str
    componentType: str
    context: str = "ekioba-marketplace"


class GenerateLayoutRequest(BaseModel):
    description: str
    context: str = "ekioba-marketplace"


class GenerateUIResponse(BaseModel):
    component: UIComponent


class GenerateLayoutResponse(BaseModel):
    layout: List[UIComponent]


if hasattr(UIComponent, "model_rebuild"):
    UIComponent.model_rebuild()
else:
    UIComponent.update_forward_refs()


@router.post("/generate-ui", response_model=GenerateUIResponse)
async def generate_ui(payload: GenerateUIRequest) -> GenerateUIResponse:
    # Placeholder generation logic; replace with model-backed generation as needed.
    component = UIComponent(
        type=payload.componentType,
        props={"className": "bg-blue-500 text-white px-4 py-2 rounded"},
        children=[UIComponent(type="span", props={"text": payload.description})],
    )
    return GenerateUIResponse(component=component)


@router.post("/generate-layout", response_model=GenerateLayoutResponse)
async def generate_layout(payload: GenerateLayoutRequest) -> GenerateLayoutResponse:
    # Example layout output: heading + button.
    layout = [
        UIComponent(type="h1", props={"text": payload.description}),
        UIComponent(
            type="button",
            props={"className": "btn-primary"},
            children=[UIComponent(type="span", props={"text": "Click Me"})],
        ),
    ]
    return GenerateLayoutResponse(layout=layout)
